modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/snapshot/IgniteClusterSnapshotCheckTest.java

```diff
@@ -584,6 +634,673 @@ public void testClusterSnapshotCheckMultipleTimes() throws Exception {
+
+        boolean requiredLeft = originatorIdx == nodeToStopIdx || grid(nodeToStopIdx).cluster().currentBaselineTopology()
+            .stream()
+            .anyMatch(bl -> bl.consistentId().equals(leaving.consistentId()));
+
+        int coordIdx = -1;
+
+        for (int i = 0; i < grids; ++i) {
+            if (stopped.contains(i))
+                continue;
+
+            final IgniteEx g = grid(i);
+
+            // Wait for all nodes complete checking.
+            assertTrue(
+                "Node " + i + " hasn't stopped the check",
+                waitForCondition(() -> !snp(g).isSnapshotChecking(SNAPSHOT_NAME), 10_000));
+
+            if (U.isLocalNodeCoordinator(g.context().discovery()))
+                coordIdx = i;
+        }
+
+        try {
+            discoSpi(grid(coordIdx)).block(msg -> msg instanceof FullMessage
+                && ((FullMessage<?>)msg).type() == CHECK_SNAPSHOT_METAS.ordinal());
+
+            IgniteInternalFuture<?> fut = snp(grid(originatorIdx)).checkSnapshot(SNAPSHOT_NAME, null, null, false, 0, true);
+
+            discoSpi(grid(coordIdx)).waitBlocked(getTestTimeout());
+
+            stopGrid(nodeToStopIdx, true);
+
+            stopped.add(nodeToStopIdx);
+
+            assertTrue(waitForCondition(() -> {
+                for (int i = 0; i < grids; ++i) {
+                    if (!stopped.contains(i) && grid(i).cluster().nodes().size() != grids - 1)
+                        return false;
+                }
+
+                return true;
+            }, getTestTimeout()));
+
+            if (nodeToStopIdx != coordIdx)
+                discoSpi(grid(coordIdx)).unblock();
+
+            if (originatorIdx == nodeToStopIdx) {
+                assertThrowsAnyCause(
+                    null,
+                    () -> fut.get(getTestTimeout()),
+                    NodeStoppingException.class,
+                    "Node is stopping"
+                );
+
+                return;
+            }
+
+            if (requiredLeft) {
+                assertThrowsAnyCause(
+                    null,
+                    () -> fut.get(getTestTimeout()),
+                    ClusterTopologyCheckedException.class,
+                    "Snapshot validation stopped. Required node has left the cluster"
+                );
+            }
+            else
+                fut.get(getTestTimeout());
+        }
+        finally {
+            if (nodeToStopIdx != coordIdx)
+                discoSpi(grid(coordIdx)).unblock();
+        }
+
+        if (requiredLeft) {
+            int chkAgainIdx = -1;
+
+            for (int i = 0; i < grids; ++i) {
+                if (stopped.contains(i))
+                    continue;
+
+                final IgniteEx g = grid(i);
+
+                // Wait for all nodes complete checking.
+                assertTrue(
+                    "Node " + i + " hasn't stopped the check",
+                    waitForCondition(() -> !snp(g).isSnapshotChecking(SNAPSHOT_NAME), 10_000));
```

Comment:
`);` on the next line or `"Node " + i + " hasn't stopped the check",` on the previous line.